### PR TITLE
GH-41112: [C++] Clean up unused parameter warnings

### PR DESCRIPTION
**Pull request by WillAyd**
Noticed these when compiling the nanoarrow test suite with -Wextra
* GitHub Issue: #41112

### Rationale for this change

Helps clean up build warnings when compiling with -Wextra

### What changes are included in this PR?

Used Arrow macros to mark some parameters as unused

### Are these changes tested?

N/A - just ensured program compiles cleanly

### Are there any user-facing changes?

No

### CODE REVIEW

**Comment by felipecrv** [comment]
I hate to be the one bringing boring news, but if a change touches code it can't be considered `MINOR: `, so please create an issue. The change looks great! Thank you for using `ARROW_UNUSED` instead of simply removing the parameter names from the function definitions.

**Comment by WillAyd** [comment]
Good to know and not a problem at all. See https://github.com/apache/arrow/issues/41112

File: cpp/src/arrow/util/decimal.h [hunk 1]
```diff
@@ -80,7 +80,7 @@ class ARROW_EXPORT Decimal128 : public BasicDecimal128 {
     std::pair<Decimal128, Decimal128> result;
     auto dstatus = BasicDecimal128::Divide(divisor, &result.first, &result.second);
     ARROW_RETURN_NOT_OK(ToArrowStatus(dstatus));
-    return std::move(result);
+    return result;
```
Comment (kou) [inline cpp/src/arrow/util/decimal.h:83]: Please don't mix unrelated changes.
These changes are for https://github.com/apache/arrow/pull/41107 , right?

File: cpp/src/arrow/array/builder_base.h [hunk 2]
```diff
@@ -177,6 +177,9 @@ class ARROW_EXPORT ArrayBuilder {
   /// The given array must be the same type as the builder.
   virtual Status AppendArraySlice(const ArraySpan& array, int64_t offset,
                                   int64_t length) {
+    ARROW_UNUSED(array);
+    ARROW_UNUSED(offset);
+    ARROW_UNUSED(length);
```
Comment (kou) [inline cpp/src/arrow/array/builder_base.h:182]: How about using `ARROW_ARG_UNUSED()` instead?

```suggestion
  virtual Status AppendArraySlice(const ArraySpan& ARROW_ARG_UNUSED(array), int64_t ARROW_ARG_UNUSED(offset),
                                  int64_t ARROW_ARG_UNUSED(length)) {
```

**Review by kou — COMMENTED** [review]
Could you use our PR template instead of removing it entirely next time?

File: cpp/src/arrow/util/decimal.h (hunk 1 above)
Comment (WillAyd) [reply cpp/src/arrow/util/decimal.h:83]: Ah yes sorry about that. Was working off of one branch to try and get a clean nanoarrow build so this slipped in. Will revert

File: cpp/src/arrow/array/builder_base.h (hunk 2 above)
Comment (WillAyd) [reply cpp/src/arrow/array/builder_base.h:182]: Sure that works. Out of curiosity now that we are on C++17 any reason not to use `[[maybe_unused]]`?

Comment (kou) [reply cpp/src/arrow/array/builder_base.h]: I think that we can use `[[maybe_unused]]`.
Could you open a new issue for this? Let's discuss this on the issue.

**Comment by pitrou** [comment]
Hmm, is this necessary? This is making the source code more annoying to read without any tangible benefit.

(we would be bound to add such annotations everywhere a stub function implementation returns `NotImplementedError`, AFAIU)

File: cpp/src/arrow/array/builder_base.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/array/builder_base.h]: Can you just try to use `[[maybe_unused]]` in this PR?

**Comment by WillAyd** [comment]
Its a fair question. It looks like gcc adds -Wunused-parameter only when both Wextra and Wall are enabled. 

I have no strong opinion either way - if we think this is too much noise in the code then I don't think its the end of the world for downstream libraries to have to ignore this warning

**Comment by felipecrv** [comment]
> Hmm, is this necessary? This is making the source code more annoying to read without any tangible benefit.
> 
> (we would be bound to add such annotations everywhere a stub function implementation returns `NotImplementedError`, AFAIU)

Then we should add `-Wno-unused-parameter` to the builds. Otherwise people that use GCC will keep sending these PRs @pitrou.

**Comment by pitrou** [comment]
@felipecrv This happens in the Arrow headers, so changing our own build flags wouldn't change anything here, AFAICT.

That said, if other people think these warnings do need to be fixed, then why not.

**Review by felipecrv — APPROVED** [review]
+1

> That said, if other people think these warnings do need to be fixed, then why not.

Yes, we should help people that build their own software with `-Werror`.

**Comment by kou** [comment]
Hmm.
It seems that `[[maybe_unused]]` doesn't work g++ 8.5.0:

https://github.com/ursacomputing/crossbow/actions/runs/8772764647/job/24072078525#step:6:167

```text
-- The CXX compiler identification is GNU 8.5.0
```

https://github.com/ursacomputing/crossbow/actions/runs/8772764647/job/24072078525#step:6:1075

```text
FAILED: src/arrow/CMakeFiles/arrow_array.dir/array/array_run_end.cc.o 
/usr/bin/c++ -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_WITH_TIMING_TESTS -I/tmp/arrow-HEAD.P3upj/cpp-build/src -I/arrow/cpp/src -I/arrow/cpp/src/generated -Wno-noexcept-type  -fdiagnostics-color=always  -Wall -fno-semantic-interposition -msse4.2  -O3 -DNDEBUG -O2 -ftree-vectorize  -std=c++17 -fPIC -MD -MT src/arrow/CMakeFiles/arrow_array.dir/array/array_run_end.cc.o -MF src/arrow/CMakeFiles/arrow_array.dir/array/array_run_end.cc.o.d -o src/arrow/CMakeFiles/arrow_array.dir/array/array_run_end.cc.o -c /arrow/cpp/src/arrow/array/array_run_end.cc
In file included from /arrow/cpp/src/arrow/array/array_run_end.cc:19:
/arrow/cpp/src/arrow/array/builder_primitive.h:38:24: error: expected unqualified-id before '[' token
   explicit NullBuilder([[maybe_unused]] const std::shared_ptr<DataType>& type,
                        ^
/arrow/cpp/src/arrow/array/builder_primitive.h:38:24: error: expected ')' before '[' token
   explicit NullBuilder([[maybe_unused]] const std::shared_ptr<DataType>& type,
                       ~^
                        )
```

**Comment by felipecrv** [comment]
> Unfortunate...so I guess back to the macros?

Yes, please.

**Comment by kou** [comment]
It seems that `[[maybe_unused]]` in constructor arguments is only unsupported:

```console
$ docker run -it --rm almalinux:8
$ dnf install -y gcc-c++ vim
$ vim a.cc
$ cat a.cc
class A {
  explicit A([[maybe_unused]] int x) {}
  void a([[maybe_unused]] int x) {}
};
int main([[maybe_unused]] int argc, [[maybe_unused]] char **argv) {
  return 0;
}
$ LANG=C g++ -std=c++17 a.cc
a.cc:2:14: error: expected unqualified-id before '[' token
   explicit A([[maybe_unused]] int x) {}
              ^
a.cc:2:14: error: expected ')' before '[' token
   explicit A([[maybe_unused]] int x) {}
             ~^
              )
```

**Comment by jonkeane** [comment]
Thanks for catching and raising that, Kou! 

I came here to say that I found similar failures in some of our extended tests for the R implementation (which relies on the C++) implementation and for historical reasons is tested across a bunch of operating systems, including some that while old are still not yet EOL. 

Which is causing out nightly builds for test-r-rstudio-r-base-4.1-opensuse153 and test-r-rstudio-r-base-4.2-centos7-devtoolset-8 as well as our r-binary packaging to fail.

**Comment by kou** [comment]
I've merged WillAyd's fix. Could you try it on your PR?

**Comment by jonkeane** [comment]
Will do! I also ran one of them on his PR too and it was good there.
